### ISSUE TITLE
set{u,g}id doesn't clear $! on success

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11649]
Migrated from [rt.perl.org#98934](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=98934) (status was 'rejected')

Searchable as RT98934$

**Comment by p5pRT** [comment]
### From cmadams@hiwaay.net

#### Created by cmadams@hiwaay.net


Setting a user/group ID by assigning to the perl special variables does
not clear $\!\, so confusing errors can occur\.  This test program fails on
5\.8\.8 \(RHEL 5\) and 5\.12\.4 \(Fedora Linux 15\)&#8203;:

\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
\#\!/usr/bin/perl

use warnings;
use strict;

my @&#8203;foo = stat \("/does/not/exist"\);
my $gid = $\( \+ 0;
$\( = $gid;
die "setgid\($gid\)&#8203;: $\!\\n" if \($\!\);
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

I get "setgid\(1000\)&#8203;: No such file or directory"\.  If I take out the
stat\(\) or add a "$\! = undef" after the stat\(\)\, the script runs as
expected\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.8 in the Red Hat build system.
It is being executed now by Perl v5.8.8 - Tue May  3 11:19:04 EDT 2011.

Site configuration information for perl v5.8.8:

Configured by Red Hat, Inc. at Tue May  3 11:19:04 EDT 2011.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.18-194.26.1.el5, archname=x86_64-linux-thread-multi
    uname='linux x86-002.build.bos.redhat.com 2.6.18-194.26.1.el5 #1 smp fri oct 29 14:21:16 edt 2010 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -Dversion=5.8.8 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64 -Dprivlib=/usr/lib/perl5/5.8.8 -Dsitelib=/usr/lib/perl5/site_perl/5.8.8 -Dvendorlib=/usr/lib/perl5/vendor_perl/5.8.8 -Darchlib=/usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi -Dsitearch=/usr/lib64/perl5/site_perl/5.8.8/x86_64-linux-thread-multi -Dvendorarch=/usr/lib64/perl5/vendor_perl/5.8.8/x86_64-linux-thread-multi -Darchname=x86_64-linux-thread-multi -Dvendorprefix=/usr -Dsiteprefix=/usr -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_en!
 dhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dinc_version_list=5.8.7 5.8.6 5.8.5 -Dscriptdir=/usr/bin'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=define uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='4.1.2 20080704 (Red Hat 4.1.2-50)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =''
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic'

Locally applied patches:
    


@INC for perl v5.8.8:
    /usr/lib64/perl5/site_perl/5.8.8/x86_64-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.8
    /usr/lib/perl5/site_perl
    /usr/lib64/perl5/vendor_perl/5.8.8/x86_64-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.8
    /usr/lib/perl5/vendor_perl
    /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi
    /usr/lib/perl5/5.8.8
    .


Environment for perl v5.8.8:
    HOME=/data/users/c/cmadams
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/data/users/c/cmadams/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @Leont

On Mon Sep 12 09&#8203;:39&#8203;:58 2011\, cmadams@&#8203;hiwaay\.net wrote&#8203;:
> \[Please enter your report here\]
> Setting a user/group ID by assigning to the perl special variables
>    does
> not clear $\!\, so confusing errors can occur\.  This test program fails
>    on
> 5\.8\.8 \(RHEL 5\) and 5\.12\.4 \(Fedora Linux 15\)&#8203;:
> 
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> \#\!/usr/bin/perl
> 
> use warnings;
> use strict;
> 
> my @&#8203;foo = stat \("/does/not/exist"\);
> my $gid = $\( \+ 0;
> $\( = $gid;
> die "setgid\($gid\)&#8203;: $\!\\n" if \($\!\);
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> 
> I get "setgid\(1000\)&#8203;: No such file or directory"\.  If I take out the
> stat\(\) or add a "$\! = undef" after the stat\(\)\, the script runs as
> expected\.

First of all\, that is exactly what you should expect\. Succeeding system
calls do not set errno\, only failing ones do\. If you want to check for
an error by using $\!\, you should set it to 0 before setting $\(\.

Secondly\, what you're doing is completely wrong\. You're setting the real
GID but not the effective GID\, thus not dropping privileges at all\. Try
POSIX' setuid/setgid\, or better yet something like Unix&#8203;::SetUser or
Proc&#8203;::UID\.

Leon Timmermans

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Chris Adams wrote&#8203;:
>Setting a user/group ID by assigning to the perl special variables does
>not clear $\!\,

Yes\, generally $\! is not cleared by successful operations\.  It is an exact
mirror for the C level errno\.  You need an independent way to see whether
the operation succeeded\, such as getting \-1 from a syscall\, and only
check $\! on failure\.  Unfortunately\, assignment to magic variables doesn't
provide a way to report failure\.  You've run into the edge of the model
here&#8203;: this is where the magic variables stop behaving like variables\.

I suggest that you use setgid\(\) from the POSIX module\.  Unfortunately it
seems not to have setegid\(\) and friends; patches welcome\.

\-zefram

**Comment by p5pRT** [comment]
### From @Leont

On Mon Sep 12 09&#8203;:39&#8203;:58 2011\, cmadams@&#8203;hiwaay\.net wrote&#8203;:
> \[Please enter your report here\]
> Setting a user/group ID by assigning to the perl special variables
> does
> not clear $\!\, so confusing errors can occur\. This test program fails
> on
> 5\.8\.8 \(RHEL 5\) and 5\.12\.4 \(Fedora Linux 15\)&#8203;:
>
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> \#\!/usr/bin/perl
>
> use warnings;
> use strict;
>
> my @&#8203;foo = stat \("/does/not/exist"\);
> my $gid = $\( \+ 0;
> $\( = $gid;
> die "setgid\($gid\)&#8203;: $\!\\n" if \($\!\);
> \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
>
> I get "setgid\(1000\)&#8203;: No such file or directory"\. If I take out the
> stat\(\) or add a "$\! = undef" after the stat\(\)\, the script runs as
> expected\.

First of all\, that is exactly what you should expect\. Succeeding system
calls do not set errno\, only failing ones do\. If you want to check for
an error by using $\!\, you should set it to 0 before setting $\(\.

Secondly\, what you're doing is completely wrong\. You're setting the real
GID but not the effective GID\, thus not dropping privileges at all\. Try
POSIX' setuid/setgid\, or better yet something like Unix&#8203;::SetUser or
Proc&#8203;::UID\.

Leon Timmermans

**Comment by p5pRT** [comment]
### From cmadams@hiwaay.net

Once upon a time\, Leon Timmermans via RT \<perlbug\-followup@&#8203;perl\.org> said&#8203;:
> On Mon Sep 12 09&#8203;:39&#8203;:58 2011\, cmadams@&#8203;hiwaay\.net wrote&#8203;:
> > \[Please enter your report here\]
> > Setting a user/group ID by assigning to the perl special variables
> >    does
> > not clear $\!\, so confusing errors can occur\.  This test program fails
> >    on
> > 5\.8\.8 \(RHEL 5\) and 5\.12\.4 \(Fedora Linux 15\)&#8203;:
> > 
> > \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> > \#\!/usr/bin/perl
> > 
> > use warnings;
> > use strict;
> > 
> > my @&#8203;foo = stat \("/does/not/exist"\);
> > my $gid = $\( \+ 0;
> > $\( = $gid;
> > die "setgid\($gid\)&#8203;: $\!\\n" if \($\!\);
> > \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
> > 
> > I get "setgid\(1000\)&#8203;: No such file or directory"\.  If I take out the
> > stat\(\) or add a "$\! = undef" after the stat\(\)\, the script runs as
> > expected\.
> 
> First of all\, that is exactly what you should expect\. Succeeding system
> calls do not set errno\, only failing ones do\. If you want to check for
> an error by using $\!\, you should set it to 0 before setting $\(\.

The problem is that is not what the documentation says\.  The docs say
"require a check to $\! to detect any possible errors"\, not "you must
clear $\! first\, make your change\, then check the results"\.

The underlying problem is that the perl source code doesn't check or
save the return code from the setregid\(\) call\.  Since library calls can
have unexpected side\-effects\, perl should handle this internally\.

For example\, my actual code that shows the problem fails after a
getpwnam\(\) for some users \(and not others\)\.  getpwnam\(\) uses the NSS
library for its lookups\, and some users are local \(/etc/passwd\) while
most are from the network \(LDAP\)\.  The NSS library tries a couple of
locations to load libnss\_ldap\.so\.2\, and errno doesn't get cleared after
the first location returns ENOENT\.

This is highly confusing\, and there's really no reason to push this off
on the people writing perl code\.  At a minimum\, perl internally could do
something like&#8203;:

&nbsp;    if \(setregid\(gid\,\-1\) == 0\)
&nbsp;        errno = 0;

\(and that goes for all the setuid/setgid family of functions\)\.

> Secondly\, what you're doing is completely wrong\. You're setting the real
> GID but not the effective GID\, thus not dropping privileges at all\. Try
> POSIX' setuid/setgid\, or better yet something like Unix&#8203;::SetUser or
> Proc&#8203;::UID\.

No\, I only want to set the real GID here \(I'm dropping privs temporarily
for a check\)\.  Also\, the above is just a short\-and\-simple example of the
problem\, not my full code\.
\-\- 
Chris Adams \<cmadams@&#8203;hiwaay\.net>
Systems and Network Administrator \- HiWAAY Internet Services
I don't speak for anybody but myself \- that's enough trouble\.

**Comment by p5pRT** [comment]
### From @Leont

On Mon\, Sep 12\, 2011 at 9&#8203;:50 PM\, Chris Adams \<cmadams@&#8203;hiwaay\.net> wrote&#8203;:
> The problem is that is not what the documentation says\.  The docs say
> "require a check to $\! to detect any possible errors"\, not "you must
> clear $\! first\, make your change\, then check the results"\.

Then the documentation should be clarified\.

> The underlying problem is that the perl source code doesn't check or
> save the return code from the setregid\(\) call\.  Since library calls can
> have unexpected side\-effects\, perl should handle this internally\.

It's an abstraction that leaks like a sieve\. If you ask me the obvious
way to deal with this would be to throw an exception on error\, but
that will probably require a «use 5\.016;»\. Using a module that does
the right dance still sounds like a better idea though\.

> This is highly confusing\, and there's really no reason to push this off
> on the people writing perl code\.  At a minimum\, perl internally could do
> something like&#8203;:
>
>    if \(setregid\(gid\,\-1\) == 0\)
>        errno = 0;
>
> \(and that goes for all the setuid/setgid family of functions\)\.

That might piss off an equal number of users because $\! gets
overwritten when they didn't expect it\.

> No\, I only want to set the real GID here \(I'm dropping privs temporarily
> for a check\)\.

Please explain yourself\. What does setting the real GID do for you? It
has very little to do with dropping any privileges\, temporarily or
permanently\.

Leon

**Comment by p5pRT** [comment]
### From @jandubois

On Tue\, 13 Sep 2011\, Leon Timmermans wrote&#8203;:
> First of all\, that is exactly what you should expect\. Succeeding
> system calls do not set errno\, only failing ones do\. If you want to
> check for an error by using $\!\, you should set it to 0 before
> setting $\(\.

That \_may\_ work\, but is not guaranteed by POSIX\.  You cannot rely
on errno not being changed by a successful call\.  The API you are
invoking may internally call other API that fail\, and set errno\,
but the outer API call expects and deals with that error condition\.

In which case you will have a successful return\, but a non\-zero errno
value\, even though you set errno to 0 before making the call\.

Cheers\,
\-Jan

**Comment by p5pRT** [comment]
### From @ikegami

On Tue\, Sep 13\, 2011 at 8&#8203;:30 AM\, Leon Timmermans \<fawaka@&#8203;gmail\.com> wrote&#8203;:

> > This is highly confusing\, and there's really no reason to push this off
> > on the people writing perl code\.  At a minimum\, perl internally could do
> > something like&#8203;:
> >
> >    if \(setregid\(gid\,\-1\) == 0\)
> >        errno = 0;
> >
> > \(and that goes for all the setuid/setgid family of functions\)\.
>
> That might piss off an equal number of users because $\! gets
> overwritten when they didn't expect it\.
>

They should be expecting it\. System calls are allowed to change $\! on
success\, since $\! isn't meaningful on success\.

**Comment by p5pRT** [comment]
### From @pjcj

On Tue\, Sep 13\, 2011 at 02&#8203;:30&#8203;:27PM \+0200\, Leon Timmermans wrote&#8203;:
> On Mon\, Sep 12\, 2011 at 9&#8203;:50 PM\, Chris Adams \<cmadams@&#8203;hiwaay\.net> wrote&#8203;:
> 
> > This is highly confusing\, and there's really no reason to push this off
> > on the people writing perl code\.  At a minimum\, perl internally could do
> > something like&#8203;:
> >
> >    if \(setregid\(gid\,\-1\) == 0\)
> >        errno = 0;
> >
> > \(and that goes for all the setuid/setgid family of functions\)\.
> 
> That might piss off an equal number of users because $\! gets
> overwritten when they didn't expect it\.

The problem in perl is that $\! is intimately bound to errno\, and you have
no business doing anything with errno except examining it immediately
after a failed system call\.  IIRC\, on some systems you cannot assign to
errno anyway since it can be implemented as a macro expanding to a
function\.

However\, this is not to say that $\! needs to remain so intimately bound to
errno\.  It's easy to make an argument that $\! should reflect the return
status of the Perl function call\, and not retain the C semantics as Jan
described\.  It's somewhat less easy to argue against that\.

If we do make such a change then it might also be worth considering the
option of throwing exceptions on failure too\.

\-\- 
Paul Johnson \- paul@&#8203;pjcj\.net
http&#8203;://www\.pjcj\.net

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed\, Sep 14\, 2011 at 01&#8203;:34&#8203;:13AM \+0200\, Paul Johnson wrote&#8203;:
> The problem in perl is that $\! is intimately bound to errno\, and you have
> no business doing anything with errno except examining it immediately
> after a failed system call\.  IIRC\, on some systems you cannot assign to
> errno anyway since it can be implemented as a macro expanding to a
> function\.

errno is required to expand to a modifiable lvalue by both c89 and
c99\, so this isn't an issue\.

Tony

**Comment by p5pRT** [comment]
### From cmadams@hiwaay.net

Once upon a time\, Leon Timmermans \<fawaka@&#8203;gmail\.com> said&#8203;:
> On Mon\, Sep 12\, 2011 at 9&#8203;:50 PM\, Chris Adams \<cmadams@&#8203;hiwaay\.net> wrote&#8203;:
> > The problem is that is not what the documentation says\.  The docs say
> > "require a check to $\! to detect any possible errors"\, not "you must
> > clear $\! first\, make your change\, then check the results"\.
> 
> Then the documentation should be clarified\.

Can you name any other part of core perl where you have to manually
reset $\! yourself?  Why should the set\{u\,g\}id changes be special?

> > The underlying problem is that the perl source code doesn't check or
> > save the return code from the setregid\(\) call\.  Since library calls can
> > have unexpected side\-effects\, perl should handle this internally\.
> 
> It's an abstraction that leaks like a sieve\. If you ask me the obvious
> way to deal with this would be to throw an exception on error\, but
> that will probably require a «use 5\.016;»\. Using a module that does
> the right dance still sounds like a better idea though\.

Using a module to paper over a deficiency in the core language is rather
backwards\.

> > This is highly confusing\, and there's really no reason to push this off
> > on the people writing perl code\.  At a minimum\, perl internally could do
> > something like&#8203;:
> >
> >    if \(setregid\(gid\,\-1\) == 0\)
> >        errno = 0;
> >
> > \(and that goes for all the setuid/setgid family of functions\)\.
> 
> That might piss off an equal number of users because $\! gets
> overwritten when they didn't expect it\.

Anybody depending on $\! to \_not\_ change is going against the
documentation already \(since the assignment is documented to affect $\!\)\.

> > No\, I only want to set the real GID here \(I'm dropping privs temporarily
> > for a check\)\.
> 
> Please explain yourself\. What does setting the real GID do for you? It
> has very little to do with dropping any privileges\, temporarily or
> permanently\.

Please stop trying to critique code you have not seen and have no idea
what it is doing\.  Here's a hint&#8203;: access\(\) \(and no\, it isn't always a
race condition\)\.

\-\- 
Chris Adams \<cmadams@&#8203;hiwaay\.net>
Systems and Network Administrator \- HiWAAY Internet Services
I don't speak for anybody but myself \- that's enough trouble\.

**Comment by p5pRT** [comment]
### From cmadams@hiwaay.net

Once upon a time\, Paul Johnson \<paul@&#8203;pjcj\.net> said&#8203;:
> On Tue\, Sep 13\, 2011 at 02&#8203;:30&#8203;:27PM \+0200\, Leon Timmermans wrote&#8203;:
> > On Mon\, Sep 12\, 2011 at 9&#8203;:50 PM\, Chris Adams \<cmadams@&#8203;hiwaay\.net> wrote&#8203;:
> > 
> > > This is highly confusing\, and there's really no reason to push this off
> > > on the people writing perl code\.  At a minimum\, perl internally could do
> > > something like&#8203;:
> > >
> > >    if \(setregid\(gid\,\-1\) == 0\)
> > >        errno = 0;
> > >
> > > \(and that goes for all the setuid/setgid family of functions\)\.
> > 
> > That might piss off an equal number of users because $\! gets
> > overwritten when they didn't expect it\.
> 
> The problem in perl is that $\! is intimately bound to errno\, and you have
> no business doing anything with errno except examining it immediately
> after a failed system call\.  IIRC\, on some systems you cannot assign to
> errno anyway since it can be implemented as a macro expanding to a
> function\.

POSIX/SuS specifcally define errno as a symbol that "shall expand to a
modifiable lvalue of type int\."  It can be a macro\, but the macro has to
expand to something that can be modified \(so not a function\)\.

Of course\, the standard also says that no standard function shall set
errno to zero\, and that it should only be examined when an error is
indicated by a function's return value\.  perl isn't bound by that of
course \(since $\! is not exactly errno; it is documented that perl
functions \_can\_ modify $\! to return their own errors\)\.

Basically\, my above code bit is a work\-around because perl calls the
set\{u\,g\}id functions on a variable assignment instead of a function call
\(that can die\)\.  The perl docs say you should check $\!\, and any sane
code would \(as they are security\-sensitive assignments\)\, and then the
only valid thing to do currently is to clear $\! first\.  IMHO if it is
required for proper functioning anyway\, the language should take care of
it for you\.

\-\- 
Chris Adams \<cmadams@&#8203;hiwaay\.net>
Systems and Network Administrator \- HiWAAY Internet Services
I don't speak for anybody but myself \- that's enough trouble\.
